### PR TITLE
Code formatting

### PR DESCRIPTION
**Pull request by louib**
Fixes https://github.com/keepassxreboot/keepassxc/issues/1168

@droidmonkey @adolfogc @TheZ3ro what would you think about this?

## How has this been tested?
```
cd build
make format
```

### CODE REVIEW

**Comment by TheZ3ro** [comment]
This will need all the code correctly formatted before merging

**Comment by louib** [comment]
@TheZ3ro we could do the formatting of all the code just before the next major, to avoid conflicts when backporting / upporting changes.

File: .travis.yml [hunk 1]
```diff
@@ -35,6 +35,7 @@ before_script:
 
 script:
   - cmake -DCMAKE_BUILD_TYPE=${CONFIG} -DWITH_GUI_TESTS=ON -DWITH_ASAN=ON -DWITH_XC_HTTP=ON -DWITH_XC_AUTOTYPE=ON -DWITH_XC_YUBIKEY=ON $CMAKE_ARGS ..
+  - make format; git diff --exit-code
```
Comment (adolfogc) [inline .travis.yml]: 👍

**Comment by adolfogc** [comment]
@louib Nice work!

**Comment by adolfogc** [comment]
> we could do the formatting of all the code just before the next major, to avoid conflicts when backporting / upporting changes.

I would like to propose the use of clang-tidy when making this one-time refactoring, as it could help fix some cases of missing readability braces around statements. Other checks could also be used, like the modernize-* ones.

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -108,9 +108,10 @@ The Branch Strategy is based on [git-flow-lite](http://nvie.com/posts/a-successf
 
 ### Coding styleguide
 
-This project follows the [Qt Coding Style](https://wiki.qt.io/Qt_Coding_Style). All submissions are expected to follow this style.
+The coding style of the project is enforced using llvm's `clang-format` formatting tool. A thorough description
+of the coding style can be found in the `.clang-format` file, but the main conventions are presented here.
```
Comment (adolfogc) [inline .github/CONTRIBUTING.md]: Would it be useful to mention here the use of `//clang-format off` and `//clang-format on` for having ClangFormat ignore certain sections of the code? To illustrate the possible use-cases, an example could also be provided. I'm thinking on something like:

When deemed necessary, you can use `//clang-format off` and `//clang-format on` to have ClangFormat ignore certain parts of the code. For instance, from project's code-base:

```cpp
// clang-format off
int binary =
    ((hmac[offset] & 0x7f) << 24)
    | ((hmac[offset + 1] & 0xff) << 16)
    | ((hmac[offset + 2] & 0xff) << 8)
    | (hmac[offset + 3] & 0xff);
// clang-format on
```

In the example above, the manual formatting resulted in more readable code than the ClangFormat changes, hence the need for using the special comment lines.

**Comment by adolfogc** [comment]
@louib After trying this branch, I have another observation:

### The steps I followed:
```bash
git clone https://github.com/louib/keepassxc.git
cd keepassxc
git checkout code_formatting
mkdir build && cd build
cmake ..
make format
```

### The error I got:
```bash
Configuration file(s) do(es) not support Objective-C: /Users/adolfogc/Downloads/keepassxc/.clang-format
make[3]: *** [CMakeFiles/format] Error 1
make[2]: *** [CMakeFiles/format.dir/all] Error 2
make[1]: *** [CMakeFiles/format.dir/rule] Error 2
make: *** [format] Error 2
```
❗️ Please note ClangFormat did affect the project files as expected.

### Possible solution:
See my other comments in the code.

File: cmake/CLangFormat.cmake [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+set(THIRD_PARTY_DIRS
```
Comment (adolfogc) [inline cmake/CLangFormat.cmake]: Include `autotype/mac` here.

File: cmake/CLangFormat.cmake [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    zxcvbn/
+    http/qhttp/
+    streams/QtIOCompressor
+)
+
+set(THIRD_PARTY_FILES
```
Comment (adolfogc) [inline cmake/CLangFormat.cmake]: Include `core/ScreenLockListenerMac.h` and `core/ScreenLockListenerMac.cpp` here.

Comment (louib) [reply cmake/CLangFormat.cmake]: Would be nice to have all the third party files centralised in one directory...

Comment (adolfogc) [reply cmake/CLangFormat.cmake]: I'm not sure those two are third party code, they need to be excluded because they contain Objective-C code.

Comment (TheZ3ro) [reply cmake/CLangFormat.cmake]: They are not third-party.

Maybe we can exclude them in another group? :thinking:

Comment (louib) [reply cmake/CLangFormat.cmake]: I just renamed the lists so that we can exclude both third-party and obj-c files.

**Comment by TheZ3ro** [comment]
I like this solution @louib

**Comment by louib** [comment]
@adolfogc I tried setting up `clang-tidy` but wasn't able to come up with a complete solution with the limited amount of time I had. Help would be appreciated, but it could be in a different PR I guess.

Can you try again `make format` on Mac, I added the obj-c files to the excluded directories.

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -113,6 +113,8 @@ of the coding style can be found in the `.clang-format` file, but the main conve
 
 Formatting can be performed automatically by calling `make format` from the `build/` directory.
 
+Note that [formatting can be disabled on a piece of code](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code) if manual formatting is deemed more readable.
```
Comment (adolfogc) [inline .github/CONTRIBUTING.md]: 👍

**Comment by adolfogc** [comment]
@louib I tested again using macOS, I can confirm the issue is now fixed.

**Comment by adolfogc** [comment]
@louib 

> I tried setting up clang-tidy but wasn't able to come up with a complete solution with the limited amount of time I had. Help would be appreciated, but it could be in a different PR I guess.

Sure, we can work on it in a next PR.

**Comment by louib** [comment]
@adolfogc cool, thanks!

If you guys are good with the solution we can keep this PR open until before the next major release and then format the codebase and merge.

I'd still like to have @droidmonkey 's approval since he wasn't convinced by the brute-force approach of enforcing formatting. My rationale for using `git diff` is that contributors can actually see in travis which sections of the code are not formatted properly and thus know that it's a formatting problem.

**Comment by phoerious** [comment]
I give my approval under the condition that we have a rough look through the code in advance and identify sections that need `// clang-format off`.

**Comment by droidmonkey** [comment]
I'm good with this approach as long as we make a massive reformat (that doesn't impact functionality) right before we complete 2.3.0

**Comment by phoerious** [comment]
After #1305 and #1327 are merged, we could do the reformatting. But this PR needs to be adjusted for TeamCity.

**Comment by louib** [comment]
@phoerious ping me when both PRs are merged, I'll do the formatting. Not sure how to do the TeamCity adjustment though.

**Comment by louib** [comment]
@keepassxreboot/core-developers added new excluded directories. I decided to exclude `src/http` since IIRC the plan is to deprecate it at some point. The style for namespaces is also causing some noise, so I guess we could decide to change it right now.

The actual formatting is in the last commit so that we can adjust the rules if needed, before merging this PR.

@phoerious where is the configuration for TeamCity? I don't see it in a file in the repo :thinking:

**Comment by phoerious** [comment]
TC has no in-repository configuration.

**Comment by phoerious** [comment]
I would suggest we do this first thing AFTER merging the release.

**Comment by phoerious** [comment]
I rebased and re-reformatted the code. Since just adding all file names to the command line crashes clang-format on my system, I changed it to use xargs instead.

**Comment by louib** [comment]
@phoerious I'll let you take care of this one.

**Comment by louib** [comment]
@phoerious rebased, changed base branch and reformatted the code.

**Comment by droidmonkey** [comment]
Do we still want to do this for 2.3.2 or 2.4 (ie develop)

**Comment by TheZ3ro** [comment]
Sincerely, I think 2.4 it's better

**Comment by TheZ3ro** [comment]
I think we can merge this, everybody agrees?

**Comment by TheZ3ro** [comment]
Seems that new code was added with #1702, we need to re-update this PR now

**Comment by louib** [comment]
@TheZ3ro should be good now.

**Pull request merged**